vms/platformvm/block/executor/block.go

```diff
@@ -29,25 +29,49 @@ func (*Block) ShouldVerifyWithContext(context.Context) (bool, error) {
 	return true, nil
 }
 
-func (b *Block) VerifyWithContext(_ context.Context, ctx *smblock.Context) error {
-	pChainHeight := uint64(0)
-	if ctx != nil {
-		pChainHeight = ctx.PChainHeight
+func (b *Block) VerifyWithContext(ctx context.Context, blockContext *smblock.Context) error {
+	var pChainHeight uint64
+	if blockContext != nil {
+		pChainHeight = blockContext.PChainHeight
 	}
 
 	blkID := b.ID()
 	if blkState, ok := b.manager.blkIDToState[blkID]; ok {
 		if !blkState.verifiedHeights.Contains(pChainHeight) {
-			// PlatformVM blocks are currently valid regardless of the ProposerVM's
-			// PChainHeight. If this changes, those validity checks should be done prior
-			// to adding [pChainHeight] to [verifiedHeights].
+			// Only the validity of warp messages need to be verified because
+			// this block has already passed verification with a different
+			// height.
+			err := VerifyWarpMessages(
+				ctx,
+				b.manager.ctx.NetworkID,
+				b.manager.ctx.ValidatorState,
+				pChainHeight,
+				b,
+			)
+			if err != nil {
+				return err
+			}
+
 			blkState.verifiedHeights.Add(pChainHeight)
 		}
 
-		// This block has already been verified.
-		return nil
+		return nil // This block has already been executed.
+	}
+
+	// Verify the warp messages in the block.
+	err := VerifyWarpMessages(
+		ctx,
+		b.manager.ctx.NetworkID,
+		b.manager.ctx.ValidatorState,
+		pChainHeight,
+		b,
+	)
+	if err != nil {
+		return err
 	}
```

Comment:
Warp messages should probably only be verified after bootstrapping is done (although technically we could verify them during bootstrapping)

Reply:
Wouldn't it be expensive to calculate the validator sets for old blocks? IMO we should get this in for this PR if I understand correctly.

Reply:
In this case, it isn't particularly expensive. On something like the C-chain, it is extremely expensive. But because the P-chain is the source of the validator sets, the referenced block height is always going to be close to the locally last accepted block.

Reply:
Makes sense, we can add a TODO in the code if we plan to add this later.

Reply:
Modified the code to handle it.